packages/common/src/sites/HubSiteManager.ts

```diff
@@ -0,0 +1,233 @@
+import { IUserRequestOptions } from "@esri/arcgis-rest-auth";
+import {
+  IHubEntityManager,
+  IHubSite,
+  IHubItemEntityManager,
+  IArcGISContext,
+  HubError,
+} from "..";
+// Node has issues if this is not directly imported
+import { ArcGISContextManager } from "../ArcGISContextManager";
+import {
+  updateSite,
+  destroySite,
+  createSite,
+  _fetchSite,
+  convertItemToSite,
+  searchSites,
+} from "./HubSites";
+import { Filter, IHubSearchOptions } from "../search";
+import { IHubRequestOptions, ISearchResponse } from "../types";
+import { setItemThumbnail as updateItemThumbnail } from "../items/setItemThumbnail";
+
+import { IItem } from "@esri/arcgis-rest-types";
+import { IRequestOptions } from "@esri/arcgis-rest-request";
+
+export class HubSiteManager
+  implements IHubEntityManager<IHubSite>, IHubItemEntityManager<IHubSite>
+{
+  /**
+   * Hold a context manager, which should be a single instance for
+   * an application. When authentication changes, the .context
+   * property on the context manager will be replaced. As long as
+   * all the fns in this class leverage properties on .context
+   * everything be kept in sync.
+   */
+  private _contextManager: ArcGISContextManager;
+  /**
+   * Hold context directly. Allows the class to be used in places
+   * where an `ArcGISContextManager` is not directly available, but
+   * an `IArcGISContext` is
+   */
+  private _context: IArcGISContext;
+
+  /**
+   * Private so we can employ a factory function should we need
+   * async work during creation
+   * @param contextOrManager
+   */
+  private constructor(contextOrManager: ArcGISContextManager | IArcGISContext) {
+    if (contextOrManager instanceof ArcGISContextManager) {
+      this._contextManager = contextOrManager;
+    } else {
+      this._context = contextOrManager;
+    }
+  }
+
+  /**
+   * Factory function to construct a new HubProjectManager instance.
+   *
+   * Note: Used so that we could do async actions in the ctor.
+   * @param contextOrManager
+   * @returns
+   */
+  static init(
+    contextOrManager: ArcGISContextManager | IArcGISContext
+  ): HubSiteManager {
+    return new HubSiteManager(contextOrManager);
+  }
+
+  /**
+   * Getter to abstract how we store the context
+   */
+  private get context() {
+    if (this._contextManager) {
+      return this._contextManager.context;
+    } else {
+      return this._context;
+    }
+  }
+
+  /**
+   * Create and store a new Site
+   *
+   * This also registers the item for oAuth and
+   * registers domain names with the hub Domain system
+   *
+   * Sites are stored as Items in the Sharing API
+   * @param site
+   * @param requestOptions
+   * @returns
+   */
+  async create(
+    site: Partial<IHubSite>,
+    requestOptions?: IHubRequestOptions
+  ): Promise<IHubSite> {
```

Comment:
Since this is a multi-step operation, I'm considering  returning the `IHubSite & OperationStack` but mulling how best to encapsulate that w/o it getting gross